### PR TITLE
LinuxRendererGLES: implement hq scalers

### PR DESCRIPTION
**Pull request by lrusak**
taken from @koying PR #12357 

This is not working properly yet. It can run with the fbo but the video does not get scaled properly, AKA. a 1280x720 video will only be 1280 pixels wide in a 1920 display. Everything including the gui isn't scaled properly so maybe that is a clue.

I haven't been able to track down the problem yet.

### CODE REVIEW

**Comment by lrusak** [comment]
hmm, this actually seems to be working ok. I think it could be an issue with GBM.

I can reset the video size by pressing `\` twice. So I think something isn't getting reset properly. Would be nice if someone could test this on another platform.

File: xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp]: no need for ifdef here, ask windowing

File: xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.cpp]: time to cleanup the ifdef hell here instead of making it worse

File: xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.cpp]: the ifdef here is a hack too

File: xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp
Comment (lrusak) [inline xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp]: unfortunately there is no `GL_EXT_framebuffer_object` in GLES so I'm not sure what to do.

File: xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.cpp]: hack, this is just an empty method for i.e. opengl

File: xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp]: the method takes a string as argument.

**Comment by pkerling** [comment]
@lrusak just cherry-picked this for wayland (pkerling/xbmc@1da938569), same problem - so it's not gbm. whole UI gets drawn at the video resolution when I switch to a HQ scaler. Switching back does not fix it.

File: xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp
Comment (koying) [inline xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp]: FBO's are core to gles 2.0, no extensions required.

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp]: then return true in windowing if this one is queried.

Comment (koying) [inline xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp]: Mmm... Returning true to a non-existing extension by if'ing an hardcoded value?
You *really* hate IFDEF's ;)

**Comment by FernetMenta** [comment]
btw: you get a much better performance if you scale the yuv image first, then do conversion to rgb.

File: xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp]: so why do you call IsSupported on the fbo if you know it is supported?

Comment (koying) [inline xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp]: https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp#L65

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/FrameBufferObject.cpp]: ahh. this check is nonsense and should be removed. opengl can test for support before calling initialize() and log a reasonable message other than "error". gles does not need to test for support.

File: xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp
Comment (koying) [inline xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp]: Looks like it requires gles3.1: https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_texture_norm16.txt

**Comment by koying** [comment]
@lrusak You miss https://github.com/xbmc/xbmc/pull/12357/files#diff-41e0ceb3de6f45669f617deae0dbeb0cR1638 for whatever reason.
That's the cause of the "corner" issue.

File: xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp
Comment (lrusak) [inline xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp]: hmm, this is included in `GLES2/gl2ext.h` from mesa. If it is a problem I can drop it.

Comment (koying) [inline xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp]: Well just check the jenkins failures ;)
Didn't dig deep, though. Just googled the symbol, and saw it's 3.1

**Comment by lrusak** [comment]
@koying thanks, that indeed fixes the issue.

File: xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp]: just leave the fbo untouched if not supported. trying to initialize on platforms that don't support it makes no sense

File: xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.h
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.h]: maybe you shoud create a new class for gles that inherits from this one.

**Comment by lrusak** [comment]
Updated, this now works properly with all scalers. This should be tested with GL also as it changes the format used for the 6x6 convolutions to use one with an alpha channel.

**Comment by FernetMenta** [comment]
before testing with openGL you should address my comments. It won't have a change to get in with those ifdefs.

**Comment by lrusak** [comment]
@FernetMenta yes of course. This is still very much a WIP. I'd like to clean it up as much as possible.

File: xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp
Comment (koying) [inline xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp]: You need `pYUVShader->SetConvertFullColorRange(!g_Windowing.UseLimitedColor());` here
The YUV shader default to limited those days...

**Comment by lrusak** [comment]
@FernetMenta I've updated to remove much of the ifdefs. I hope that is better!

File: xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp
Comment (lrusak) [inline xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp]: :+1:

**Comment by lrusak** [comment]
jenkins build this please

File: xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.h
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.h]: this is not a good idea. I consider OpenGL reference and don't want ports like GLES to break when the reference changes.

File: xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.h
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.h]: I don't like this. That's kind of the same hack the fist GLES rendering fo vaapi had shown.

File: xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.cpp]: the diff between OpenGL and GLES gets too big for keeping both implementations in the same file

**Comment by lrusak** [comment]
I'm not going to split the file. I think that is counterproductive. If OpenGL breaks OpenGLES then it will get fixed. If it was split then OpenGLES will be forgotten about.

I'm not happy about the VAAPI change but it was the only way forward. The `RendererVAAPIGL.cpp` and `RendererVAAPIGLES.cpp` are the exact same, there is no reason for them to be split other than the fact that that's how you wanted them to be.

In my opinion the path forward is to make the OpenGL code compatible with OpenGLES. This should be done by only using OpenGLES compatible calls.

**Comment by FernetMenta** [comment]
Without splitting the file this won't get merged. No discussion about this. You can do by yourself or wait until I'll find time to do so. The latter will delay this PR.

File: xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp]: no vertical alignment

File: xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp]: put every assignment on a separate line

File: xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.cpp]: no verticl alignment

File: xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.cpp]: shozld be done in header, not in constructor (C++11 style)

File: xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.cpp]: no vertial alignment

**Comment by lrusak** [comment]
Pushed some fixes so gl works like gles for the videofiltershader. I have tested and it is working.

@FernetMenta I addressed some of your concerns and will fix the rest soon.

**Comment by lrusak** [comment]
Rebased, should make for an easier review

File: xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp]: that looks wrong. we render from fbo that is loaded into texture units. what is your intent?

File: xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp]: keep GL_QUADS or provide performance measures that justifiy the change

**Comment by FernetMenta** [comment]
if you want to speed this up, drop all OpenGL related changes and focus only on what the title of this PR says.

File: xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.h
Comment (garbear) [inline xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.h]: clang's sanitizer will strip leading virtuals, so the next time someone does this dropping virtual will make their lives easier

File: xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp
Comment (lrusak) [reply xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp]: It is present in the GLES renderer however it doesn't seem to effect anything with or without it so I removed it now.

File: xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp
Comment (lrusak) [reply xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp]: I am not sure how to measure the performance here. I can only provide some links:

https://stackoverflow.com/questions/14555446/opengl-is-it-more-efficient-to-use-gl-quads-or-gl-triangles
https://stackoverflow.com/questions/6644099/what-is-so-bad-about-gl-quads

**Comment by lrusak** [comment]
The changes to GL are only so I don't have to ifdef the hell out of everything.

File: xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.h
Comment (lrusak) [reply xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShader.h]: thanks, I don't know how that ended up there in the first place :)

**Comment by FernetMenta** [comment]
this won't get enywhere if you touch GL. I insist to keep GL_QUADS as it is used in the other render methods. Create a separate videofilter for GLES as I already suggested.

**Comment by lrusak** [comment]
```
lukas@lukas-macbook-pro ~/Documents/git/xbmc$ find xbmc/ -type f | xargs grep GL_QUADS 
xbmc/pictures/SlideShowPicture.cpp:  glBegin(GL_QUADS);
xbmc/guilib/GUIFontTTFGL.cpp:  glDrawArrays(GL_QUADS, 0, m_vertex.size());
xbmc/guilib/GUITextureGL.cpp:  glBegin(GL_QUADS);
xbmc/guilib/GUITextureGL.cpp:  glBegin(GL_QUADS);
xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp:  glBegin(GL_QUADS);
xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp:  glBegin(GL_QUADS);
xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp:  glBegin(GL_QUADS);
xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp:  glBegin(GL_QUADS);
xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp:  glBegin(GL_QUADS);
xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererGL.cpp:  glDrawArrays(GL_QUADS, 0, m_count * 4);
xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererGL.cpp:  glBegin(GL_QUADS);
```

I really would rather avoid code duplication again. I mean look at the vaapi example. http://sprunge.us/DMCD

**Comment by FernetMenta** [comment]
this is the way to go, I want to keep this separately. some duplicate lines of code it not an issue. different maintainers, different files.

**Comment by lrusak** [comment]
Updated to conform to @FernetMenta 's requests.

**Review by FernetMenta — APPROVED** [review]

**Comment by lrusak** [comment]
jenkins build this please

File: xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShaderGLES.cpp
Comment (Rechi) [inline xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShaderGLES.cpp]: fails on RBPI beacause `GL_RGBA16F_EXT` isn't declared

Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShaderGLES.cpp]: RBPI builds should not include this file

Comment (lrusak) [reply xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShaderGLES.cpp]: why not?

Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/VideoFilterShaderGLES.cpp]: it only uses its own renderer

**Comment by lrusak** [comment]
I'm assuming android, rbpi, aml, and imx don't use `LinuxRendererGLES` can anyone confirm?

**Comment by FernetMenta** [comment]
android does use it: https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererMediaCodec.h#L29

**Comment by koying** [comment]
> I'm assuming android, rbpi, aml, and imx don't use LinuxRendererGLES can anyone confirm?

Well, all GLES platforms use it in ffmpeg/software mode, really (besides maybe rpi)

**Comment by lrusak** [comment]
>Well, all GLES platforms use it in ffmpeg/software mode, really (besides maybe rpi)

You're right, if I disable mediacodec on android I can use the HQ scaler methods.

I've updated to not build only on RPi

File: xbmc/cores/VideoPlayer/VideoRenderers/CMakeLists.txt
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/CMakeLists.txt]: actually this does not depend on whether GLES was found or not. it depends on if the platform depends of GLES renderer or not. means the logic here is not correct.

Comment (lrusak) [reply xbmc/cores/VideoPlayer/VideoRenderers/CMakeLists.txt]: Ok I revised it to just depend on which platforms need it

File: xbmc/cores/VideoPlayer/VideoRenderers/CMakeLists.txt
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/CMakeLists.txt]: wayland can use OpenGL or GLES

Comment (lrusak) [reply xbmc/cores/VideoPlayer/VideoRenderers/CMakeLists.txt]: I really don't know what you want. In this case I should leave the OPENGLES_FOUND statement,

Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/VideoRenderers/CMakeLists.txt]: I just point out (once again) that our build system does not fit our requirements. The current issues you encounter are a clear indicator for this. 

Ideally a platform maintainer should be able to include platform dependent code in the build. The platform maintainer is the only one how really knows that.

**Comment by lrusak** [comment]
Updated with the best possible solution that our build system provides.

**Comment by lrusak** [comment]
Rebased and squashed. Would be nice if I could get an ok from other GLES platform maintainers. ping @koying @Memphiz @peak3d

**Comment by Memphiz** [comment]
I have not the slightest amount of time for testing this currently. But given the fact that the last test definitely didn't break anything, and I am pretty sure this is still the case in the current state of this PR - you have my OK ;)

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
jenkins build this please

**Comment by Rechi** [comment]
jenkins build this please

File: xbmc/cores/VideoPlayer/VideoRenderers/CMakeLists.txt
Comment (Rechi) [inline xbmc/cores/VideoPlayer/VideoRenderers/CMakeLists.txt]: use `CORE_PLATFORM_NAME_LC` instead of `CORE_PLATFORM_NAME`
see comment at https://github.com/xbmc/xbmc/blob/4dfc7a8/cmake/scripts/common/Platform.cmake#L13-L16

Comment (lrusak) [reply xbmc/cores/VideoPlayer/VideoRenderers/CMakeLists.txt]: ah right, thanks

**Comment by lrusak** [comment]
I don't see any reason why rbpi should be failing, it should not even be building that code

**Comment by lrusak** [comment]
squshed, rbpi built successfully. Ready to merge :+1:

**Comment by Memphiz** [comment]
Thx for getting it through :)

**Pull request merged**
